blockchain/transaction.go

```diff
@@ -1,52 +1,60 @@
 package blockchain
 
-import (
-	"crypto/sha256"
-	"encoding/binary"
+import "encoding/binary"
+
+const (
+	// TxHashPointerLen is the length in bytes of a hash pointer.
+	TxHashPointerLen = 32/8 + HashLen
+	// TxOutputLen is the length in bytes of a transaction output.
+	TxOutputLen = 64/8 + AddrLen
 )
 
 // TxHashPointer is a reference to a transaction on the blockchain.
 type TxHashPointer struct {
-	blockNumber uint32
-	hash        Hash
+	BlockNumber uint32
+	Hash        Hash
 }
 
 // Marshal converts a TxHashPointer to a byte slice
 func (thp TxHashPointer) Marshal() []byte {
-	buf := []byte{}
-	binary.LittleEndian.PutUint32(buf, thp.blockNumber)
-	for _, b := range thp.hash {
-		buf = append(buf, b)
-	}
+	buf := make([]byte, 4, TxHashPointerLen)
+	binary.LittleEndian.PutUint32(buf, thp.BlockNumber)
+	buf = append(buf, thp.Hash.Marshal()...)
 	return buf
 }
 
 // TxOutput defines an output to a transaction
 type TxOutput struct {
-	amount    uint64
-	recipient Wallet
+	Amount    uint64
+	Recipient Address
 }
 
 // Marshal converts a TxOutput to a byte slice
 func (to TxOutput) Marshal() []byte {
-	buf := []byte{}
-	binary.LittleEndian.PutUint64(buf, to.amount)
-	buf = append(buf, to.recipient.Marshal()...)
+	buf := make([]byte, 8, TxOutputLen)
+	binary.LittleEndian.PutUint64(buf, to.Amount)
+	buf = append(buf, to.Recipient.Marshal()...)
 	return buf
 }
 
 // TxBody contains all relevant information about a transaction
 type TxBody struct {
-	sender  Wallet
-	input   TxHashPointer
-	outputs []TxOutput
+	Sender  Address
+	Input   TxHashPointer
+	Outputs []TxOutput
+}
+
+// Len returns the length of a transaction body
+func (tb TxBody) Len() int {
```

Comment:
Perhaps `func (tb *TxBody) Len() int` would be better here? Not sure about the semantics, but something to think about.